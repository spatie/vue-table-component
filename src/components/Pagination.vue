<template>
    <nav v-if="shouldShowPagination">
        <ul class="pagination justify-content-center">
            <li :class="{ 'disabled': pagination.currentPage === 1 }">
                <a href="javascript:;" aria-label="First" @click="pageClicked(1)">
                    <span aria-hidden="true" v-html="first"></span>
                </a>
            </li>
            <li :class="{ 'disabled': pagination.currentPage === 1 }">
                <a href="javascript:;" aria-label="Previous" @click="pageClicked(pagination.currentPage - 1)">
                    <span aria-hidden="true" v-html="previous"></span>
                </a>
            </li>
            <li
                    v-for="page in pagesLeft"
                    class="page-item"
                    :class="{ 'active': pagination.currentPage === page }"
            >
                <a href="javascript:;" @click="pageClicked(page)">{{ page }}</a>
            </li>

            <li
                    v-if="pagination.lastPage > 10 && pagesMiddle && pagesMiddle.length"
                    class="page-item"
            >
                <a>...</a>
            </li>
            <li
                    v-if="pagination.lastPage > 10 && pagesMiddle && pagesMiddle.length"
                    v-for="page in pagesMiddle"
                    class="page-item"
                    :class="{ 'active': pagination.currentPage === page }"
            >
                <a href="javascript:;" @click="pageClicked(page)">{{ page }}</a>
            </li>
            <li
                    v-if="pagination.lastPage > 10 && pagesMiddle && pagesMiddle.length"
                    class="page-item"
            >
                <a>...</a>
            </li>

            <li
                    v-if="pagination.lastPage > 10 && !pagesMiddle"
                    class="page-item"
            >
                <a>...</a>
            </li>

            <li
                    v-if="pagesRight.length"
                    v-for="page in pagesRight"
                    class="page-item"
                    :class="{ 'active': pagination.currentPage === page }"
            >
                <a href="javascript:;" @click="pageClicked(page)">{{ page }}</a>
            </li>
            <li :class="{ 'disabled': pagination.currentPage === pagination.lastPage }">
                <a href="javascript:;" aria-label="Next" @click="pageClicked(pagination.currentPage + 1)">
                    <span aria-hidden="true" v-html="next"></span>
                </a>
            </li>
            <li :class="{ 'disabled': pagination.currentPage === pagination.lastPage }">
                <a href="javascript:;" aria-label="Last" @click="pageClicked(pagination.lastPage)">
                    <span aria-hidden="true" v-html="last"></span>
                </a>
            </li>
        </ul>
    </nav>
</template>

<script>
    import range from 'lodash/range';

    export default {
        props: {
            pagination: {
                type: Object,
                default: () => ({}),
            },
        },

        data(){
            return {
                pagesLeft: [],
                pagesMiddle: [],
                pagesRight: [],
                next: '&raquo;',
                last: '&raquo;&raquo;',
                previous: '&laquo;',
                first: '&laquo;&laquo;',
            }
        },

        computed: {
            shouldShowPagination() {
                if (this.pagination.lastPage === undefined) {
                    return false;
                }

                this.pages(this.pagination.currentPage);

                return this.pagination.perPage < this.pagination.total;
            },
        },

        methods: {
            isActive(page) {
                const currentPage = this.pagination.currentPage || 1;

                return currentPage === page;
            },
            pages(page) {
                if(this.pagination.lastPage + 1 <= 10) {
                    this.pagesLeft = range(1, this.pagination.lastPage + 1);
                    this.pagesRight = [];
                    this.pagesMiddle = null;
                }
                else if(page <= 6){
                    this.pagesLeft = range(1, 8);
                    this.pagesRight = range(this.pagination.lastPage - 2, this.pagination.lastPage + 1);
                    this.pagesMiddle = null;
                }
                else if( (page > (this.pagination.lastPage - 4)) || (page  === this.pagination.lastPage + 1) ){
                    this.pagesLeft = range(1, 3);
                    this.pagesMiddle = null;
                    this.pagesRight = range(this.pagination.lastPage - 4, this.pagination.lastPage + 1);
                }
                else {
                    this.pagesLeft = range(1, 3);
                    this.pagesMiddle = range(page - 2, page + 3);
                    this.pagesRight = range(this.pagination.lastPage - 1, this.pagination.lastPage + 1);
                }
            },
            pageClicked(page) {
                if (this.pagination.currentPage === page) {
                    return;
                }

                if(page < 1 || page > this.pagination.lastPage) {
                    return;
                }

                this.$emit('pageChange', page);

                this.pages(page);
            },
        },
        watch: {
            pagination() {
                this.pages(this.pagination.currentPage);
            }
        }
    };
</script>
