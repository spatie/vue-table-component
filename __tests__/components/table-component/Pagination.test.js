import Pagination from '../../../src';
import Vue from 'vue/dist/vue.js';


describe('Pagination', () => {
    Vue.use(Pagination);

    it('can mount without pagination data', async () => {
        document.body.innerHTML = `
            <div id="app">
                <pagination></pagination>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will not display when there is only one page', async () => {

        document.body.innerHTML = `
            <div id="app">
                <pagination
                :pagination="{
                    currentPage: 1,
                    lastPage: 1,
                    total: 1,
                }">

                </pagination>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('will render links when there is more than one page', async () => {

        document.body.innerHTML = `
            <div id="app">
                <pagination
                :pagination="{
                    currentPage: 1,
                    lastPage: 3,
                    perPage: 10,
                    total: 30,
                }">

                </pagination>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('can set the active page', async () => {

        document.body.innerHTML = `
            <div id="app">
                <pagination
                :pagination="{
                    currentPage: 2,
                    lastPage: 3,
                    perPage: 10,
                    total: 30,
                }">

                </pagination>
            </div>
        `;

        await createVm();

        expect(document.body.innerHTML).toMatchSnapshot();
    });
});

async function createVm() {
    const vm = new Vue({
        el: '#app',
    });

    await Vue.nextTick(() => {
    });

    const table = vm.$children[0];

    return table;
}
