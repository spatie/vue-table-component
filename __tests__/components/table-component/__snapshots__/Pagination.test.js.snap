// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination can mount without pagination data 1`] = `

<div id="app">
  <!---->
</div>

`;

exports[`Pagination can set the active page 1`] = `

<div id="app">
  <nav>
    <ul class="pagination justify-content-center">
      <li class>
        <a href="javascript:;"
           aria-label="First"
        >
          <span aria-hidden="true">
            ««
          </span>
        </a>
      </li>
      <li class>
        <a href="javascript:;"
           aria-label="Previous"
        >
          <span aria-hidden="true">
            «
          </span>
        </a>
      </li>
      <li class="page-item">
        <a href="javascript:;">
          1
        </a>
      </li>
      <li class="page-item active">
        <a href="javascript:;">
          2
        </a>
      </li>
      <li class="page-item">
        <a href="javascript:;">
          3
        </a>
      </li>
      <!---->
      <!---->
      <!---->
      <li class>
        <a href="javascript:;"
           aria-label="Next"
        >
          <span aria-hidden="true">
            »
          </span>
        </a>
      </li>
      <li class>
        <a href="javascript:;"
           aria-label="Last"
        >
          <span aria-hidden="true">
            »»
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>

`;

exports[`Pagination will not display when there is only one page 1`] = `

<div id="app">
  <!---->
</div>

`;

exports[`Pagination will render links when there is more than one page 1`] = `

<div id="app">
  <nav>
    <ul class="pagination justify-content-center">
      <li class="disabled">
        <a href="javascript:;"
           aria-label="First"
        >
          <span aria-hidden="true">
            ««
          </span>
        </a>
      </li>
      <li class="disabled">
        <a href="javascript:;"
           aria-label="Previous"
        >
          <span aria-hidden="true">
            «
          </span>
        </a>
      </li>
      <li class="page-item active">
        <a href="javascript:;">
          1
        </a>
      </li>
      <li class="page-item">
        <a href="javascript:;">
          2
        </a>
      </li>
      <li class="page-item">
        <a href="javascript:;">
          3
        </a>
      </li>
      <!---->
      <!---->
      <!---->
      <li class>
        <a href="javascript:;"
           aria-label="Next"
        >
          <span aria-hidden="true">
            »
          </span>
        </a>
      </li>
      <li class>
        <a href="javascript:;"
           aria-label="Last"
        >
          <span aria-hidden="true">
            »»
          </span>
        </a>
      </li>
    </ul>
  </nav>
</div>

`;
